pkg/progressreader/progressreader_test.go

```diff
@@ -0,0 +1,79 @@
+package progressreader
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"io/ioutil"
+	"testing"
+
+	"github.com/docker/docker/pkg/streamformatter"
+)
+
+func TestOutputOnPrematureClose(t *testing.T) {
+	var outBuf bytes.Buffer
+	content := []byte("TESTING")
+	reader := ioutil.NopCloser(bytes.NewReader(content))
+	writer := bufio.NewWriter(&outBuf)
+
+	prCfg := Config{
+		In:        reader,
+		Out:       writer,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      len(content),
+		NewLines:  true,
+		ID:        "Test",
+		Action:    "Read",
+	}
+	pr := New(prCfg)
+
+	part := make([]byte, 4, 4)
+	_, err := io.ReadFull(pr, part)
+	if err != nil {
+		pr.Close()
+		t.Fatal(err)
+	}
+
+	writer.Flush()
+	tlen := outBuf.Len()
+	pr.Close()
+	writer.Flush()
+	if outBuf.Len() == tlen {
+		t.Fatalf("Expected some output when closing prematurely")
+	}
+}
+
+func TestCompleteSilently(t *testing.T) {
+	var outBuf bytes.Buffer
+	content := []byte("TESTING")
+	reader := ioutil.NopCloser(bytes.NewReader(content))
+	writer := bufio.NewWriter(&outBuf)
+
+	prCfg := Config{
+		In:        reader,
+		Out:       writer,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      len(content),
+		NewLines:  true,
+		ID:        "Test",
+		Action:    "Read",
+	}
+	pr := New(prCfg)
+
+	out, err := ioutil.ReadAll(pr)
+	if err != nil {
+		pr.Close()
```

Comment:
It's 'undefined' according to http://golang.org/pkg/io/#Closer. I didn't use `defer` in this function precisely because there must be an explicit call to Close() at the end (necessary for the test), and the Close() in this case has some non-trivial logic in it. I would say that it's better to call Close() explicitly on all exit points (I missed one after line 68).